### PR TITLE
LOOP-887 Keycloak update and DIY Sync

### PR DESCRIPTION
**Pull request by ps2**
Keycloak changes ticket - https://tidepool.atlassian.net/browse/LOOP-887
DIY Sync ticket - https://tidepool.atlassian.net/browse/LOOP-4610

Reviewers Note - this is a large ticket, but not as large as the 18K loc change number might suggest; most of those lines are translation updates.

The Keycloak support depends on some changes to LoopKit, including the PluginHost protocol and updates to the ServiceDelegate protocol to allow issuing of alerts when a token is revoked or no longer valid.

Coupled with these changes are other changes in DIY that are fairly extensive, like re-arranging files to allow LoopKit to be built as a Swift Package. It is possible to decouple these changes, and submit them to Tidepool repos separately, but that leads to a diverged set of forks that may make later incorporation of these changes conflict and error prone, so I am initially proposing we take these changes together.

Here is the list of included changes, along with links to the DIY tickets or PRs that should provide the rationale. Any changes that impact Tidepool Loop functionality should be documented in associated Jira tickets.

* [Fix Loop-dev crash from Correction Range accepted by Onboarding](https://github.com/LoopKit/LoopKit/pull/436)
* [LoopKit iOS target bumped to 15.1, and Support plugins can now vend configuration menu items](https://github.com/LoopKit/LoopKit/pull/444)
* [Services can record “identity” values for analytics](https://github.com/LoopKit/LoopKit/commit/c1a9842e5d7c2ab13fca803143faf7a5bde2d6a8)
* [Add unique index to dosing decision store](https://github.com/LoopKit/LoopKit/pull/445)
* [Include mutable doses from InsulinDeliveryStore in normalized doses](https://github.com/LoopKit/LoopKit/pull/446)
* [Expose hostIdentifier and hostVersion to service plugins](https://github.com/LoopKit/LoopKit/commit/1ccc830b1884b928b04f320738c3bb3d1f51cdb5)
* Updated translations
* [Store override history without conflicting overrides](https://github.com/LoopKit/LoopKit/pull/449)
* [Fixes for HealthKit unresponsiveness and Loop crash](https://github.com/LoopKit/LoopKit/pull/452)
* [Add missed meal notifications to Loop](https://github.com/LoopKit/LoopKit/pull/428)
* [Update max bolus information screen](https://github.com/LoopKit/LoopKit/pull/442)
* [Types for remote commands](https://github.com/LoopKit/LoopKit/pull/453)
* [Added food emojis](https://github.com/LoopKit/LoopKit/pull/458)
* [Updates for tracking automatic flag on doses](https://github.com/LoopKit/LoopKit/commit/87900b8cbe09792f664961c2253f82634f58dfd6)

### CODE REVIEW

File: LoopKit Example/MasterViewController.swift [hunk 1]
```diff
@@ -206,7 +214,8 @@ class MasterViewController: UITableViewController {
         case .data:
             switch DataRow(rawValue: indexPath.row)! {
             case .carbs:
-                performSegue(withIdentifier: CarbEntryTableViewController.className, sender: sender)
+                //performSegue(withIdentifier: CarbEntryTableViewController.className, sender: sender)
```
Comment (Camji55) [inline LoopKit Example/MasterViewController.swift:217]: Still needed?

File: LoopKit/CarbKit/CachedCarbObject+CoreDataClass.swift [hunk 2]
```diff
@@ -139,11 +138,11 @@ extension CachedCarbObject {
...
-        precondition(!object.createdByCurrentApp)
-        precondition(sample.createdByCurrentApp == object.createdByCurrentApp)
-        precondition(sample.provenanceIdentifier == object.provenanceIdentifier)
-        precondition(object.syncIdentifier != nil)
-        precondition(sample.syncIdentifier == object.syncIdentifier)
+//        precondition(!object.createdByCurrentApp)
+//        precondition(sample.createdByCurrentApp == object.createdByCurrentApp)
+//        precondition(sample.provenanceIdentifier == object.provenanceIdentifier)
+//        precondition(object.syncIdentifier != nil)
+//        precondition(sample.syncIdentifier == object.syncIdentifier)
```
Comment (Camji55) [inline LoopKit/CarbKit/CachedCarbObject+CoreDataClass.swift]: Needed?

File: LoopKit/CarbKit/CarbStore.swift [hunk 3]
```diff
@@ -206,7 +206,8 @@ public final class CarbStore: HealthKitSampleStore {
         calculationDelta: TimeInterval = 5 /* minutes */ * 60,
         effectDelay: TimeInterval = 10 /* minutes */ * 60,
         carbAbsorptionModel: CarbAbsorptionModel = .nonlinear,
-        provenanceIdentifier: String
+        provenanceIdentifier: String,
+        test_currentDate: Date? = nil
```
Comment (Camji55) [inline LoopKit/CarbKit/CarbStore.swift:210]:
```suggestion
        testCurrentDate: Date? = nil
```

Any reason for the mixed casing here?

**Review by Camji55 — APPROVED** [review]

File: LoopKit/InsulinKit/DoseStore.swift [hunk 4]
```diff
@@ -201,6 +202,7 @@ public final class DoseStore {
     ///   - syncVersion: A version number for determining resolution in de-duplication
     ///   - lastPumpEventsReconciliation: The date the PumpManger last reconciled with the pump
     ///   - provenanceIdentifier: An id to store with new doses, indicating the provenance of the dose, usually the app's bundle identifier.
+    ///   - readyCallback: A closure that will be called after initialization.
```
Comment (nhamming) [inline LoopKit/InsulinKit/DoseStore.swift]: [nit] align with parameter name `onReady`.

File: LoopKitUI/ServiceUI.swift [hunk 5]
```diff
@@ -30,7 +30,7 @@ public protocol ServiceUI: Service {
     /// - Parameters:
     ///     - colorPalette: Color palette to use for any UI.
     /// - Returns: Either a conforming view controller to create and onboard the service or a newly created and onboarded service.
-    static func setupViewController(colorPalette: LoopUIColorPalette) -> SetupUIResult<ServiceViewController, ServiceUI>
+    static func setupViewController(colorPalette: LoopUIColorPalette, pluginHost: PluginHost) -> SetupUIResult<ServiceViewController, ServiceUI>
```
Comment (nhamming) [inline LoopKitUI/ServiceUI.swift:34]: Update comment ^ to include this new parameter

File: LoopKitUI/SupportUI.swift [hunk 6]
```diff
@@ -33,15 +33,24 @@ public protocol SupportUI: AnyObject {
     /// - Returns: A view that will be used in a support menu for providing user support.
     func supportMenuItem(supportInfoProvider: SupportInfoProvider, urlHandler: @escaping (URL) -> Void) -> AnyView?
 
+    /// Provides configuration menu items.
+    ///
+    /// - Parameters:
+    ///   - supportInfoProvider: A provider of additional support information.
+    ///   - urlHandler: A handler to open any URLs.
+    /// - Returns: An array of views that will be added to the configuration section of settings.
+    func configurationMenuItems() -> [AnyView]
```
Comment (nhamming) [inline LoopKitUI/SupportUI.swift:49]: there are no parameters in the function. update comment

File: LoopKit/GlucoseRangeSchedule+SafeBounds.swift [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+import HealthKit
+
+extension GlucoseRangeSchedule {
+    public func safeSchedule(with suspendThreshold: HKQuantity?) -> GlucoseRangeSchedule? {
+        let minGlucoseValue = [
+            suspendThreshold?.doubleValue(for: self.unit),
+            Guardrail.correctionRange.absoluteBounds.lowerBound.doubleValue(for: self.unit)
+        ]
+            .compactMap({ $0 })
+            .max()!
```
Comment (nhamming) [inline LoopKit/GlucoseRangeSchedule+SafeBounds.swift:19]: guard instead of force unwrap?

File: LoopKit/HealthKitSampleStore.swift [hunk 8]
```diff
@@ -48,15 +47,19 @@ public class HealthKitSampleStore {
     /// For unit testing only.
     internal var testQueryStore: HKSampleQueryTestable?
 
+    /// Current date. Will return the unit-test configured date if set, or the current date otherwise.
+    internal var currentDate: Date {
+        test_currentDate ?? Date()
```
Comment (nhamming) [inline LoopKit/HealthKitSampleStore.swift:52]: camel case?

File: LoopKit/HealthKitSampleStore.swift [hunk 9]
```diff
@@ -157,98 +157,180 @@ public class HealthKitSampleStore {
...
-        // Subclasses can override
+    private struct QueryState: Equatable {
+        var anchorState: QueryAnchorState
+        var authorizationDetermined: Bool
+
+        // ObserverQuery will fail if HKAuthorizationStatus is undetermined
+        var readyToQuery: Bool {
+            return anchorState != .uninitialized && authorizationDetermined
+        }
+    }
```
Comment (nhamming) [inline LoopKit/HealthKitSampleStore.swift:204]: [general comment] are these changes covered by unit tests? If not, please add some.

**Review by nhamming — APPROVED** [review]
If this update could benefit from additional unit testing, please add.

File: LoopKit Example/MasterViewController.swift (hunk 1 above)
Comment (ps2) [reply LoopKit Example/MasterViewController.swift:217]: No, the example app for LoopKit is in a pretty sad, unused state.  We could maybe remove the whole thing.

File: LoopKit/CarbKit/CachedCarbObject+CoreDataClass.swift (hunk 2 above)
Comment (ps2) [reply LoopKit/CarbKit/CachedCarbObject+CoreDataClass.swift]: No

File: LoopKit/CarbKit/CarbStore.swift (hunk 3 above)
Comment (ps2) [reply LoopKit/CarbKit/CarbStore.swift:210]: To signify it's "special", i.e. not part of the normal interface, but used for testing. It follows the convention set on the other stores.

File: LoopKit/GlucoseRangeSchedule+SafeBounds.swift (hunk 7 above)
Comment (ps2) [reply LoopKit/GlucoseRangeSchedule+SafeBounds.swift:19]: We know the last two elements of minGlucoseValue are non-nil, so I think this is ok.

File: LoopKit/HealthKitSampleStore.swift (hunk 8 above)
Comment (ps2) [reply LoopKit/HealthKitSampleStore.swift:52]: This is a special naming convention that was started on the other stores, to denote a parameter that is "special", and not part of the normal API.

File: LoopKit/HealthKitSampleStore.swift (hunk 9 above)
Comment (ps2) [reply LoopKit/HealthKitSampleStore.swift:204]: Yes, this code is covered in a number of tests that exercise store initialization with health kit authorization.

**Pull request merged**